### PR TITLE
[extension/storage/filestorage] Default fsync to true

### PR DESCRIPTION
**Pull request by bouk**
**Description:**

See #31200, setting fsync to false by default is dangerous and will lead to data loss.

**Link to tracking Issue:** #31200

### CODE REVIEW

**Review by djaglowski — COMMENTED** [review]
Although the component is in beta, I am in support of this change because it makes the default configuration much safer. However, I would like to hear other opinions as well. cc: @open-telemetry/collector-contrib-approvers 

If we do make the change, we'll need to roll it out with a feature gate. ([Example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e1403742f39d090bfc3c164f5a295d1783df3fb1/extension/storage/filestorage/extension.go#L19))

**Comment by astencel-sumo** [comment]
> If we do make the change, we'll need to roll it out with a feature gate.

I fully agree this is a breaking change, but I'm not convinced that flipping a feature gate is easier/more feasible than changing the configuration value?

**Comment by djaglowski** [comment]
> I fully agree this is a breaking change, but I'm not convinced that flipping a feature gate is easier/more feasible than changing the configuration value?

I think the benefit is to give users warning that the default will change, but maybe it's just more hassle then it's worth.

**Review by andrzej-stencel — APPROVED** [review]

**Comment by swiatekm-sumo** [comment]
I'd like to see some benchmarks before we merge this. The storage extension is used by the persistent queue in core, which writes each request to the queue. Enabling or disabling fsync - which in this case only protects against hardware failure and similar events - should be a decision the user makes consciously, as it has potentially major performance implications depending on the storage medium in question.

**Comment by bouk** [comment]
@swiatekm-sumo you can see some benchmarks on the original file storage PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3087

**Comment by djaglowski** [comment]
> @swiatekm-sumo you can see some benchmarks on the original file storage PR: #3087

Based on [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3087#issuecomment-820878692), we observed a 3x difference in performance.

We had a longer discussion about the risks of this setting on that thread as well. Based on rereading, it's not clear to me that this change should happen. Perhaps we should consider a third option which is to add more detailed documentation about the parameter _and remove_ the default value, forcing users to make a decision about it.

I will say that in the absence of a clear consensus I think we need to default to no breaking change.

**Review by djaglowski — CHANGES_REQUESTED** [review]
I'm blocking this until we establish consensus. I'll call attention to this discussion at next week's SIG meeting. 

cc: @open-telemetry/collector-contrib-approvers in case there are additional opinions.

**Comment by swiatekm-sumo** [comment]
Part of the problem is that the value of this setting should likely depend on what the extension is used for. The persistent queue use case is performance-sensitive, but use cases where we just store some state, like in filelog receiver, would probably be better off with fsync on.

There's also some effort to make the persistent queue less taxing on the storage, so once that lands we could revisit whether enabling fsync remains painful there.

**Comment by bouk** [comment]
It's a shame that bbolt panics on corruption instead of allowing a graceful way to truncate the database

**Comment by djaglowski** [comment]
@swiatekm-sumo, I'd be interested to see your findings on other storage engines. I seem to recall writing up a comparison vs other options when first implementing the extension but can't seem to find it now.

**Comment by swiatekm-sumo** [comment]
@djaglowski I'll see if I can dig it up. Even if I can't, I don't think it'll be difficult to reproduce.

**Comment by djaglowski** [comment]
It was pointed out in SIG today that the performance impact of enabling fsync was original stated as "3 orders of magnitude", rather than the 3x I stated above. 

In any case, we agreed that given the tradeoffs here, we should serious consider switching away from bbolt. I suggest we should wait until @swiatekm-sumo has a chance to look for his previous analysis. Otherwise, if anyone has potential alternatives in mind, please feel free to call them out.

**Pull request closed without merging**
